<a href="https://colab.research.google.com/github/valesierrai/Analitica-de-Negocios/blob/main/Taller2_%C3%81rbolesdeDecisi%C3%B3n.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

0. Cargamos Librerias

In [ ]:
#Conexión con el Drive
from google.colab import drive
drive.mount('/content/drive')

import numpy as np
import pandas as pd
from sklearn.metrics import confusion_matrix
from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import export_graphviz
from pydotplus import graph_from_dot_data

Mounted at /content/drive


1. Cargamos los Datos

In [ ]:
nxl='/content/drive/MyDrive/Colab Notebooks/Analítica de Negocios/2. Taller 2 - Datos_Conitnue_Edu.xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)
XDB=XDB[['Age','Marriage','Income','Edu','Own','Pet','City','ContinueEdu']]
XDB=XDB.dropna()

#Configuración de la matriz de entrada y del vector de datos deseado
XD=np.array(XDB[['Age','Marriage','Income','Edu','Own','Pet','City']])
yd=np.array(XDB[['ContinueEdu']])

2. Implementación del modelo

In [ ]:
mar=DecisionTreeClassifier(criterion='gini', max_depth=4)
mar.fit(XD,yd)

#Se evalua el modelo
ydp=mar.predict(XD)
cm=confusion_matrix(yd,ydp)

array([[1789,    5],
       [  51,  155]])

3. Árbol de Desición

In [ ]:
vs=['Age','Marriage','Income','Edu','Own','Pet','City'] #Titulos del Árbol
dot_data=export_graphviz(mar,feature_names=vs) #Coloque los nombres al árbol
graph=graph_from_dot_data(dot_data) #Crear el gráfico
graph.write_png('Arbol_EduContinua.png')

True

4. Evaluación del modelo (métricas)

In [ ]:
VP=cm[0,0];FN=cm[0,1];FP=cm[1,0];VN=cm[1,1]
Ex=(VP+VN)/(VP+VN+FP+FN)
print("La exactitud es:", Ex)

Te=(FN+FP)/(VP+VN+FP+FN)
print("La tasa de error es:", Te)

Pr=VP/(VP+FP)
print("La precisión es:", Pr)

Esp=VN/(FN+VN)
print("La espicificidad es:", Esp)

La exactitud es: 0.972
La tasa de error es: 0.028
La precisión es: 0.9722826086956522
La espicificidad es: 0.96875


5. Análisis de Resultados

Del proceso de los datos, se puede observar que el número de personas interesadas es de 206 y de no interesadas de 1794. Teniendo porcentajes de 10.30% (206/2000) y 89.70%(1794/2000) respectivamente. La ejecución del modelo dio como resultado cuatro nodos puros en el árbol. Los cuales están distribuidos en el nivel 3 nodo 8 ([0,138]), y en el nivel 4 los nodos 2 ([1059,0]), 5([0,3]), y el nodo 14([4,0]). Con base a esto, se podrá encontrar la regla de negocio perteneciente a cada uno de estos nodos. En este caso analizaremos la regla de negocio para el nodo con mayor numero de registros (nivel 4 nodo 2 [1059,0])sería Income <=136500.0, Income <=111000.0, income <= 99000.0, y Marriage >10.5; y la regla para el nodo puro con menor número de registros (nivel 4 nodo 5 [0,3]) es Income <=136500.0, Income >111000.0, Pet <=0.5, y City<=0.5. Asimismo, encontramos que la regla de negocio Income>136500 and Edu<=1.5 and Pet<=0.5 and Income>179,000, pertenece al nodo con los intervalos [2, 8] con gini 0.32. Así como esta Income<=136,500 and Income >111,000 and Pet>=0.5 and Edu>=1.5 pertence al nodo [50, 11] con gini 0.296. Por último, la matriz de confusión realizada para este ejercicio nos proporcionó diferentes métricas para el análisis del modelo. La exactitud alcanzada por el modelo es de 97.20%, teniendo una tasa de error de 2.80%. Además, se econtró el nivel de precisión para el modelo que arrojó un 97.23% y para la espicificidad del pronóstico dió un 96.88%.